Always set relevant variables for cross compiling

bnrjee opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
1. Right now, there seems to be no way to override the platform specific cc flags that kubernetes sets. We should be able to override those values.
For example, bottlerocket sets customized CC flags for different platforms using this [patch](https://github.com/bottlerocket-os/bottlerocket/blob/41a72fe21400fa227543fcbed2e59eb0a8b633c1/packages/kubernetes-1.17/0001-always-set-relevant-variables-for-cross-compiling.patch). We should be able to override the default values in kubernetes using environment variables or build time flags and build kubernetes with customized gcc variants that bottlerocket uses. 
2. We want to be able to build kubernetes on host platforms other than amd64. For example, we may want to build amd64 targets on arm platforms. For that to happen, we need to remove the condition that checks if the build platform is amd64. If no platform is specified, we can use default values.

**Which issue(s) this PR fixes**:
Fixes #94291

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md


-->
```
```
**Release note**:
```release-note
Allow cross compilation of kubernetes on different platforms.
```

bnrjee committed:
Always set relevant variables for cross compiling

Based on a patch to Bottlerocket by Ben Cressey <bcressey@amazon.com>
https://github.com/bottlerocket-os/bottlerocket/blob/41a72fe21400fa227543fcbed2e59eb0a8b633c1/packages/kubernetes-1.17/0001-always-set-relevant-variables-for-cross-compiling.patch

micahhausler commented:
/sig release

bnrjee commented:
/assign @spiffxp

hwdef commented:
/ok-to-test

bnrjee commented:
@juanvallejo, @SataQiu , would it be possible to add this to milestone v1.20?

spiffxp commented:
/assign @BenTheElder 
This LGTM but I'd like a second pair of eyes for the /approve
/milestone v1.20
/lgtm

bnrjee commented:
Hi @BenTheElder, 
Would it be possible for you to take a look at this?

BenTheElder commented on hack/lib/golang.sh:
```diff
@@ -392,30 +392,36 @@ kube::golang::set_platform_envs() {
-        export CC=s390x-linux-gnu-gcc
-        ;;
-    esac
+  # Apply standard values for CGO_ENABLED and CC unless KUBE_BUILD_PLATFORMS is set.
```
why?

bnrjee replied:
So, the idea here is to use default values, if we do not want cross compilation.

BenTheElder replied:
we should just set them as default without the branch then and override them when cross compiling.
this is not an accurate check for cross-compilation.

BenTheElder commented on hack/lib/golang.sh:
```diff
@@ -392,30 +392,36 @@ kube::golang::set_platform_envs() {
+      return
   fi
+  # Dynamic CGO linking for other server architectures goes here
+  # If you want to include support for more server platforms than these, add arch-specific gcc names here
```
this is no longer guarded by being on a supported build host.

bnrjee replied:
We want to build kubernetes on platforms other than amd64. Updated the motivation statement to reflect that.

BenTheElder replied:
ok, but that's not possible with this PR alone, so we're no longer correctly checking ...?

BenTheElder reviewed (commented):
This change does not seem to gel with the motivation comment in the PR body.

If the motivation is being able to override the CC flags, we should just support _overriding_ them instead of altering how they're set to begin with.
I don't think the switch from checking the host platform to checking if KUBE_BUILD_PLATFORMS is set makes sense.

BenTheElder commented:
/cc @mkumatag @dims

bnrjee commented:
> This change does not seem to gel with the motivation comment in the PR body.
> 
> If the motivation is being able to override the CC flags, we should just support _overriding_ them instead of altering how they're set to begin with.
> I don't think the switch from checking the host platform to checking if KUBE_BUILD_PLATFORMS is set makes sense.
Hi @BenTheElder
Sorry about the confusion. I have edited the motivation statement in PR body. We want to override the cc values. We also want to be able to build kubernetes on platforms other than amd64.

bnrjee commented:
Hi  @BenTheElder,
I have updated the motivation statement. Would it be possible for you to take another look at these code changes?

BenTheElder reviewed (commented):
please remove the check for support from the upstream code once it's actually supported to cross build from another platform
the upstream kube-cross / kube build image is only amd64 currently, when that is fixed we can possibly support building from another host.

BenTheElder commented:
cc @kubernetes/release-engineering @kubernetes/sig-release 
cc @dims <-- recently led some discussions around architecture support

BenTheElder commented on hack/lib/golang.sh:
```diff
@@ -392,30 +392,36 @@ kube::golang::set_platform_envs() {
+  case "${platform}" in
+    "linux/amd64")
+      export CGO_ENABLED=1
+      export CC=${LINUX_AMD64_CC:-gcc}
```
this change is fine. though usually we guard non-standard env with a `KUBE_` prefix

dims commented:
i'll wait for @BenTheElder to be happy :)

bnrjee committed:
Support cross compilation only on amd64.

Based on a patch to Bottlerocket by Ben Cressey <bcressey@amazon.com>
https://github.com/bottlerocket-os/bottlerocket/blob/41a72fe21400fa227543fcbed2e59eb0a8b633c1/packages/kubernetes-1.17/0001-always-set-relevant-variables-for-cross-compiling.patch

bnrjee commented:
/test pull-kubernetes-node-e2e

dims commented:
/approve
/lgtm
